addons/website_sale/controllers/main.py

```diff
@@ -249,10 +249,52 @@ def shop(self, page=0, category=None, search='', ppg=False, **post):
         Product = request.env['product.template'].with_context(bin_size=True)
 
         search_product = Product.search(domain, order=self._get_search_order(post))
+        product_count = len(search_product)
+
         website_domain = request.website.website_domain()
         categs_domain = [('parent_id', '=', False)] + website_domain
         if search:
-            search_categories = Category.search([('product_tmpl_ids', 'in', search_product.ids)] + website_domain).parents_and_self
+            if product_count < request.env.cr.IN_MAX:
```

Comment:
Could be a condition depending on System Parameter instead
```suggestion
            if not request.env['ir.config_parameter'].get_param('website_sale.products_optimization', False):
```